WallsAndHoles/tilematerial.h

```diff
@@ -21,32 +19,51 @@ class TileMaterial : public QObject
     /**
      * @brief Creates a TileMaterial with the given material.
      */
-    TileMaterial(QSharedPointer<QImage> texture, float ambient, float diffuse, float specular, float shininess, QObject *parent = nullptr);
-
-
-    QSharedPointer<QImage> getTexture() const;
-    float getAmbient() const;
-    float getDiffuse() const;
-    float getSpecular() const;
-    float getShininess() const;
-
-
-    void setTexture(QSharedPointer<QImage> texture);
+    TileMaterial(QString name,
+                 SharedImageAndSource texture,
+                 float ambient,
+                 float diffuse,
+                 float specular,
+                 float shininess,
+                 QObject *parent = nullptr);
+
+    QString name() const;
+    SharedImageAndSource texture() const;
+    float ambient() const;
+    float diffuse() const;
+    float specular() const;
+    float shininess() const;
+
+
+    void setName(QString name);
+    void setTexture(SharedImageAndSource texture);
     void setAmbient(float ambient);
     void setDiffuse(float diffuse);
     void setSpecular(float specular);
     void setShininess(float shininess);
 
     static TileMaterial *getDefaultMaterial();
 
-signals:
+    bool operator ==(const TileMaterial &other)
+    {
+        return mAmbient == other.mAmbient
+                && mDiffuse == other.mDiffuse
+                && mSpecular == other.mSpecular
+                && mShininess == other.mShininess
+                && mTexture == other.mTexture
```

Comment:
Might even be good to implement operator==() in ImageAndSource to just compare source files, and replace this line with
```
&& *mTexture == *(other.mTexture)
```